### PR TITLE
Colour 'Labels' layers by cell statistics

### LINKED ISSUE #51: Add option to colour the segmentation by cell statistics
Currently the regions in the segmented images are coloured by region label. It would be nice if the regions could instead be coloured by cell statistic (e.g. area, perimeter, number of neighbours, etc.)

We can pass [a dictionary of {`region_id: colour`}](https://napari.org/stable/api/napari.layers.Labels.html#napari.layers.Labels.color) to the labels layer to specify the colour for each region. Not sure if there's a better way to do it though. How would we let the user set the lower / upper limits for the colourmap?

### PR DESCRIPTION
**Pull request by p-j-smith**
Fixes #51 

https://user-images.githubusercontent.com/29753790/230074652-1bf638b0-7e98-47fa-8288-bbf9f8479aef.mp4

- set minimum version of `scikit-image>=0.20`, as part of the API that we're using change in this version (`skimage.future.RAG` was renamed to `skimage.graph.RAG`). Also remove unused dependencies

- add widgets for creating colourmaps based on cell statistics

- to create a colourmap: select a `Labels` layer to create the colourmaps for, set the range to use for creating the colourmaps, then click 'Create colourmaps'

- this will create a series of colourmaps for the selected statistic - one per frame - and store them as a list in `Labels.metadata["colourmaps"]`

- when scrolling through frames, the colourmap is updated for the current frame

- current options for colouring by statistic:
    - id - in which case the `Labels.color_mode` is set to `auto`, which will colour the cells by id
    - area
    - perimeter
    - orientation
    - number of neighbours

- there is an `Auto-limits` option for setting the range for the colourmaps. If selected, the minimum and maximum values of the selected statistic (across all frames) will be used for setting the range. When creating the colourmap, the 'lower limit' and 'upper limit' widgets will be updated with the automatically calculated values so the user know what values have been used

- there are some issues / discussion with colouring a `Labels` layer by features in `napari` that we might to want keep track of:
    - https://github.com/napari/napari/issues/4384
    - https://github.com/napari/napari/issues/5129
    - https://napari.zulipchat.com/#narrow/stream/212875-general/topic/labels.20.28direct.29
    - https://forum.image.sc/t/exloring-image-segment-features-in-napari/75222/3

- there are some other plugins that create colourmaps based on cell statistics, but I couldn't find any that allow setting the colourmap range. Plus, they tend to do the `regionprops` calculation on-the-fly when the frame is changed rather than storing results for all frames in the metadata, which means there is usually a lag when scrolling through frames while the data is processed and colourmaps created

### CODE REVIEW

File: src/napari_epitools/_widget.py [hunk 1]
```diff
@@ -369,11 +384,97 @@ def _create_cell_statistics_widgets() -> list[Widget]:
...
+    )
+
+    statisitc_tooltip = "Select which statistic to use for colouring the 'Labels'"
+    statistic = magicgui.widgets.create_widget(
+        name="colourmap_statistic",
+        value="None",
+        label="statistic",
+        widget_type="ComboBox",
+        options={
+            "choices": ["None", "area", "perimeter", "orientation", "neighbours"],
```
Comment (paddyroddy) [inline src/napari_epitools/_widget.py:411]: I feel these should all be capitalised or none of them

File: src/napari_epitools/_widget.py [hunk 2]
```diff
@@ -466,3 +576,87 @@ def _cell_statistics_to_csv(
...
+    statistics: npt.NDArray,
+    indices: npt.NDArray,
+) -> dict[int, npt.NDArray]:
+    """Create a dictionary of colours - one per unique region"""
+
+    colourmap_data = (statistics - lower_limit) / upper_limit
+    colourmap_data = np.clip(colourmap_data, a_min=0, a_max=1)
+    colours = plt.cm.turbo(colourmap_data)
+
+    return {index: colour for index, colour in zip(indices, colours)}
```
Comment (paddyroddy) [inline src/napari_epitools/_widget.py:662]:
```suggestion
    return dict(zip(indices, colours))
```

File: src/napari_epitools/_widget.py [hunk 3]
```diff
@@ -369,11 +384,97 @@ def _create_cell_statistics_widgets() -> list[Widget]:
...
+    lower_limit_tooltip = "Set the lower limit for the colourmap"
+    lower_limit = magicgui.widgets.create_widget(
+        value=0.0,
+        name="colourmap_lower_limit",
+        label="lower limit",
+        widget_type="FloatSpinBox",
+        options={
+            "tooltip": lower_limit_tooltip,
+            "min": -3.4 * 10**38,
+            "max": 3.4 * 10**38,
```
Comment (paddyroddy) [inline src/napari_epitools/_widget.py:425]: Any reason for 10^38?

**Review by paddyroddy — APPROVED** [review]
A couple of minor comments.

Looks really cool 🥳 looks like Pablo liked it too!

File: src/napari_epitools/_widget.py (hunk 1 above)
Comment (p-j-smith) [reply src/napari_epitools/_widget.py:411]: yeah good point - I'll change "None" to "id" as that's actually what it does

File: src/napari_epitools/_widget.py (hunk 3 above)
Comment (p-j-smith) [reply src/napari_epitools/_widget.py:425]: by default `magicgui` sets an upper limit of 1000 to the float widgets (I'm not sure what the lower limit is). Some of the values in the statistics were larger than this, but I didn't really know what a reasonable upper limit would be for them. I [checked the maximum value](https://en.wikipedia.org/wiki/Single-precision_floating-point_format#:~:text=floating%2Dpoint%20variable%20has%20a%20maximum%20value%20of%20(2%20%E2%88%92%202%E2%88%9223)%20%C3%97%202127%20%E2%89%88%203.4028235%20%C3%97%201038) a `float32` can take whilst keeping [reasonable precision](https://en.wikipedia.org/wiki/Single-precision_floating-point_format#:~:text=This%20gives%20from%206%20to%209%20significant%20decimal%20digits%20precision.), and apparently it's roughly `3.4 x 10^38`!

Comment (paddyroddy) [reply src/napari_epitools/_widget.py:425]: Fair enough haha!

**Pull request merged**